### ISSUE TITLE
Hyperopt and overfitting (discussion)

### ISSUE THREAD

**Issue by gaugau3000** [freqtrade/freqtrade#2472]
> “I remember my friend Johnny von Neumann used to say, with four
> parameters I can fit an elephant, and with five I can make him wiggle his
> trunk.” [Enrico Fermi, 1953]

As a end user and newbie to algo trading i prefer to see negative profit when optimize instead of see positive but not reproducible results and think i have find the holly grail (everybody have the feeling once ).

# the roi
in hyperopt we can optimize roi but i think the way it is design it lead to overfitting :
the backtest profits should follow a kind of normal distribution.

![image](https://user-images.githubusercontent.com/2643583/68124054-b69edb80-ff0e-11e9-968b-e281119f4695.png)

But with hyperopt you can easily introduce some "peaks" in the distribution that fit well in in-sample but leads to loose money in out-sample. Why ? because you have maybe 10 very good trades with a lot of profit that you have find only because of brute force but you never find this trades in the future.

# the stop loss

Do you trade a strategy with 5 minutes time-frame who look likes this after optimization :

> stop_loss = -0.35 
> minimum_roi = {'0' : 0.02, '60':0}

Not sure. Why ? because your risk reward is terrific. Maybe the optimization give you this result but what happens in live after 3 loose trades.

Solution : pretty sure you can design loss function to prevent this behavior but maybe it will take more iterations to converge. Can we limit over-fitting in search spaces to prevent some optimized unbalance strategies.

What do you think about his ?

**Comment by xmatthias** [comment]
I think this is where a good loss-function comes in, which avoids overfitting.

Using the `OnlyProfitHyperOptLoss` - you'll most likely see this behaviour (that's why i don't really like this loss-function), unless your `'hyperopt_min_trades'` is well adapted your timerange (it'll strongly vary if you hyperopt a week or a year).

SharpeLoss should somewhat cover this... as it takes multiple dimensions into account in calculating the loss.

As all of Hyperopt, loss-functions rely on `'hyperopt_min_trades'` - which in hindsight should better be a "trades per day" variable - not an absolute one - to make it independent of the timerange - since having 30 trades in a week can be good - but having 30 trades in a year is closer to HODL ... 

Another aspect to this is the timerange - if you hyperopt 5m timeframe with a 1week timerange, it'll be overfitted to the situation of this week.
Ideally, longer timeranges should somewhat cover this - always assuming that the loss-function balances well between profit, trades,  +/- trade ratio, ( ... ) - and isn't one-dimensionally optimizing on profits - or trade duration - or ( ... )


Yes, a loss-function that prevents this will take longer to converge - but if the results are better that's a tradeoff i'll take any day.

**Comment by djacky** [comment]
I agree with @xmatthias...Also, I suggest to perform a _Walk Forward Analysis_ in your backtests to confirm your optimization and strategy...Do a Google search on this and you will find a lot of useful information.

**Comment by gaugau3000** [comment]
Hi,

Honestly i think it is very difficult to prevent over-fitting no matter the loss function we use (i have perform lots of testing with different loss functions with no real success : outliers detection, linear regression, standard deviations, calmar ratio, Risk/Reward... with different weights). Hyper-opt is a kind of super grid-search when you have many parameters as input for machine learning problem. But the loss function is compute comparing a test set and training set. In freqtrade we use hyperopt  to find best paramaters again and again one the same datas : of course we will suffer of the Survivorship Bias.

@djacky tks for response. I already perform walk foward analysis and i am pretty sure that it is a feature that has to be implement in hyperopt to see how viable the strategy is.

An the search spaces i think should limit to find for a needle in a haystack. For example stop loss 0.01 step (I know it is not possible to define a step for now as it is integer or float as dimensions).

I think there is no magic if you have an average strategy hyperopt will not turn into incredible one even worse he can easily overfit it if you don't have a minimum of knowledge.

I don't use anymore hyperopt or only to perform a walk foward analysis on only 1 param (with steps) -> so it is more a kind of gridSearch :-) 

Have a nice day.

**Comment by djacky** [comment]
I'm not sure if Hyperopt actually does Walk Forward analysis...I think they just minimize an objective function subject to constraints. I agree that it is difficult to find good solutions with Hyperopt...this is mainly due to the fact that the optimization problems are highly nonlinear, and there may be many local minima to the problem. If the problem was convex, the best solution could have been found in seconds. The market is, unfortunately, very nonlinear and unpredictable, and so we have to live with that...

**Comment by xmatthias** [comment]
I'm going to close this as i don't think there's much we can do to prevent overfitting in hyperopt, except by analyzing parameters with "human understanding" (use 0.05 as stoploss, not 0.051231) ...

**Comment by Danson77** [comment]
One way to fix overfitting is punishing the model for Hyperopt with code like         

try:
            # Ensure trade_count is more than 2 for meaningful calculations
            if trade_count < 2:
                sharpe_ratio = 1
                sortino_ratio = 1
                calmar_ratio = 1

            # If in profit
            if profit > 0:

                  loss = (
                   Your Metric
            )

            # If not in profit < 0
            else:
                loss = profit

            # Final return
            return -loss


But when using     else: loss = profit    It will throw you just okay hyper parameters, I don't use it I go straight for best-performing hyperopt one space by space not all at once, Make sure you remove pairs you don't want the parameters to fit in to, like 
leverages ".*(2L|2S|3L|3S|5L|5S|UP|DOWN)/.*" or 
FIAT  "(AUD|BRZ|CAD|CHF|EUR|GBP|HKD|IDRT|JPY|NGN|RUB|SGD|TRY|UAH|USD|ZAR)/.*" or 
stable coins "(BUSD|CUSD|CUSDT|DAI|PAXG|SUSD|TUSD|USDC|USDN|USDP|USDT|VAI|UST|USDD|USDJ|USTC|AUSD|OUSD)/.*"

Using all the up to date active pairs is recommended, once you hyperopt parameters spaces enough 
"I dot two runs per space, 1 month last one mostly, 8 workers because even with 32 cores and 128gb ram still kills workers with error if above 8 (876 pairs minus the banned ones ) 1000  EPOCH takes me • 5:07:04 H, so double that per space "
Then Backtest with last known good hyperparameters and Blacklist Bad Pairs that taking you down,you can keep repeating until  you got zero or close to non of the drownown pairs if you want to overfit just a little because after all you don't want to keep bad resulting pairs even if in the future they might give us some profit which usually is too small to be worth it I just let other strategy be fitted to that pair instead,, now your parameters are ready and pairlist is ready to put in to live or if its your first time run dry run for the night if you unsure about your Loss function, 

I use my own and done lots of testing in controlled environment so I know it does improve parameters well,but if you using standard default models like 

 ProfitDrawDownHyperOptLoss will show me 42USDT+ - 4.7 DROWDOWN 78.6% winrate
SortinoHyperOptLoss with 36USDT + - 4.4 DROWNODOWN 70.8% winrate

those are aim for the high profits but have terrible drowdown,

SharpeHyperOptLoss will come as 35 USD which is terrible profit and only 2.7 drowdown is better 73.1% winrate

Non of them hit higher then avg profit 1.20% and my aim now is above +1.60 avg profit with +77% winrate, that's why its important to make your own metrics and mathematical question to use hyperopt efficiently and do a lot of controlled environment testings and you will see how the hyperopt models perform this will remove a lot of guessing and the unknown that kept me stuck, at the end I recommend implementing own metric model for hyperopt its simple yet so effective and it was a good learning curve after learning writing strategies
